CesiumGltf/src/Model.cpp

```diff
@@ -402,76 +402,84 @@ void forEachPrimitiveInMeshObject(
-      0.0,
-      1.0};
+std::optional<glm::dmat4x4> getNodeTransform(const CesiumGltf::Node& node) {
+  if (!node.matrix.empty() && node.matrix.size() < 16) {
+    return std::nullopt;
+  }
 
-  // This should just call GltfUtilities::getNodeTransform, but it can't because
-  // it's in CesiumGltfContent. We should merge these two libraries, but until
-  // then, it's duplicated.
+  // clang-format off
+  // This is column-major, just like glm and glTF
+  static constexpr std::array<double, 16> identityMatrix = {
+      1.0, 0.0, 0.0, 0.0,
+      0.0, 1.0, 0.0, 0.0,
+      0.0, 0.0, 1.0, 0.0,
+      0.0, 0.0, 0.0, 1.0};
+  // clang-format on
 
-  glm::dmat4x4 nodeTransform = transform;
   const std::vector<double>& matrix = node.matrix;
 
-  if (node.matrix.size() == 16 &&
-      !std::equal(matrix.begin(), matrix.end(), identityMatrix.begin())) {
-
-    const glm::dmat4& nodeTransformGltf =
-        *reinterpret_cast<const glm::dmat4*>(matrix.data());
+  if (node.matrix.size() >= 16 && !std::equal(
+                                      identityMatrix.begin(),
+                                      identityMatrix.end(),
+                                      matrix.begin())) {
+    return glm::dmat4x4(
+        glm::dvec4(matrix[0], matrix[1], matrix[2], matrix[3]),
+        glm::dvec4(matrix[4], matrix[5], matrix[6], matrix[7]),
+        glm::dvec4(matrix[8], matrix[9], matrix[10], matrix[11]),
+        glm::dvec4(matrix[12], matrix[13], matrix[14], matrix[15]));
+  }
 
-    nodeTransform = nodeTransform * nodeTransformGltf;
-  } else if (
-      !node.translation.empty() || !node.rotation.empty() ||
+  if (!node.translation.empty() || !node.rotation.empty() ||
       !node.scale.empty()) {
-
-    glm::dmat4 translation(1.0);
-    if (node.translation.size() == 3) {
+    glm::dmat4x4 translation(1.0);
+    if (node.translation.size() >= 3) {
       translation[3] = glm::dvec4(
           node.translation[0],
           node.translation[1],
           node.translation[2],
           1.0);
+    } else if (!node.translation.empty()) {
+      return std::nullopt;
     }
 
     glm::dquat rotationQuat(1.0, 0.0, 0.0, 0.0);
-    if (node.rotation.size() == 4) {
+    if (node.rotation.size() >= 4) {
       rotationQuat[0] = node.rotation[0];
       rotationQuat[1] = node.rotation[1];
       rotationQuat[2] = node.rotation[2];
       rotationQuat[3] = node.rotation[3];
+    } else if (!node.rotation.empty()) {
+      return std::nullopt;
     }
 
-    glm::dmat4 scale(1.0);
-    if (node.scale.size() == 3) {
+    glm::dmat4x4 scale(1.0);
+    if (node.scale.size() >= 3) {
       scale[0].x = node.scale[0];
       scale[1].y = node.scale[1];
       scale[2].z = node.scale[2];
+    } else if (!node.scale.empty()) {
+      return std::nullopt;
     }
 
-    nodeTransform =
-        nodeTransform * translation * glm::dmat4(rotationQuat) * scale;
+    return translation * glm::dmat4x4(rotationQuat) * scale;
   }
 
+  return glm::dmat4x4(1.0);
+}
```

Comment:
This looks like a lot of changes, but really all I did was duplicate `GltfUtilities::getNodeTransform` into here. The previous code did the same thing but with some small differences.

Also, this code used to be inside `forEachPrimitiveInNodeObject` but now it's factored out into a helper function that `forEachNode` can also use.